### PR TITLE
Add optional support for SonarQube Integration

### PR DESCRIPTION
**Pull request by tsukhu**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Most CI/CD systems will have integration to continous code quality systems namely SonarCube. This PR adds SonarCube integration for Angular 4 projects .

* **What is the current behavior?** (You can also link to an open issue here)

N/A

* **What is the new behavior (if this is a feature change)?**

Optional installation notes for SonarCube integration have been added . This feature will push code quality matrix to SonarCube based on linting , coverage etc.

* **Other information**:
Here are some sample screenshots

Angular Starter project dashboard
sonarcube1

Quality Metrics
sonarcube2

A simulated lint issue added to view results on SonarCube
sonarcube3_error_simulation

### CODE REVIEW

**Comment by tsukhu** [comment]
Added support for unit test results as well
sonarcube4

**Comment by gdi2290** [comment]
this is really kool

**Comment by gdi2290** [comment]
can you redo the PR? and make it optional?

**Comment by tsukhu** [comment]
@gdi2290 sure .. will take the latest version and make that change ..

**Comment by tsukhu** [comment]
While I am at it found another issue related to a corejs warning (https://github.com/angular/angular/issues/20357)

We get this error during the tests 
WARNING in ./node_modules/@angular/core/esm5/core.js
6581:15-36 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/@angular/core/esm5/core.js
 @ ./src/app/app.component.spec.ts
 @ ./src \.spec\.ts
 @ ./config/spec-bundle.js

WARNING in ./node_modules/@angular/core/esm5/core.js
6601:15-102 Critical dependency: the request of a dependency is an expression
 @ ./node_modules/@angular/core/esm5/core.js
 @ ./src/app/app.component.spec.ts
 @ ./src \.spec\.ts
 @ ./config/spec-bundle.js

Solution:

https://github.com/gdi2290/angular-starter/commit/268f54ae499eb1727f6ca326bce1b769c45f7873

**Comment by tsukhu** [comment]
One more point .. Can we add   browserNoActivityTimeout: 30000 in the karma.conf.js .. (increase from the default to 3x the value) 

For slower machines we do run into this issue at times ...

**Comment by tsukhu** [comment]
@gdi2290 I have made the required changes for making SonarQube optional. We need to use npm test:sonar for generating the sonar reports. Also updated the README for the same.

File: README.md [hunk 1]
```diff
@@ -225,6 +226,28 @@ npm run e2e:live
 npm run build:docker
 ```
 
+### Optional Integration with SonarQube (for continous code quality)
```
Comment (gdi2290) [inline README.md]: can you move this after Netlify

File: config/karma.conf.js [hunk 2]
```diff
@@ -60,8 +60,18 @@ module.exports = function (config) {
       type: 'in-memory'
     },
 
+    sonarQubeUnitReporter: {
```
Comment (gdi2290) [inline config/karma.conf.js]: can you move this to process.env.SONAR_QUBE

File: config/karma.conf.js [hunk 3]
```diff
@@ -60,8 +60,18 @@ module.exports = function (config) {
...
+      outputFile: 'reports/ut_report.xml',
+      overrideTestDescription: true,
+      testPath: 'src/app',
+      testFilePattern: '.spec.ts',
+      useBrowserName: false
+    },
+
     remapCoverageReporter: {
       'text-summary': null,
+      lcovonly: './coverage/coverage.lcov',
```
Comment (gdi2290) [inline config/karma.conf.js]: can you move this to process.env.SONAR_QUBE unless this is for something else

File: package.json [hunk 4]
```diff
@@ -134,6 +135,7 @@
     "karma-jasmine": "^1.1.0",
     "karma-mocha-reporter": "^2.2.3",
     "karma-remap-coverage": "^0.1.4",
+    "karma-sonarqube-unit-reporter": "0.0.13",
```
Comment (gdi2290) [inline package.json]: can you move this to the readme so it's not downloaded then check if it's installed when you run karma

File: sonar-project.properties [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+sonar.ts.tslint.configPath=tslint.json
+sonar.ts.coverage.lcovReportPath=coverage/coverage.lcov
+# if using local tslint then enable the line below
+# sonar.ts.tslint.outputPath=reports/lint_issues.json
+sonar.genericcoverage.unitTestReportPaths=reports/ut_report.xml
+
+# Change the host.url to point to the 
+# sonarcube server (default localhost)
+sonar.host.url=http://localhost:9000
```
Comment (gdi2290) [inline sonar-project.properties]: can you put this in the readme

File: README.md (hunk 1 above)
Comment (tsukhu) [reply README.md]: Done

File: config/karma.conf.js (hunk 2 above)
Comment (tsukhu) [reply config/karma.conf.js]: Done

File: config/karma.conf.js (hunk 3 above)
Comment (tsukhu) [reply config/karma.conf.js]: Done

File: package.json (hunk 4 above)
Comment (tsukhu) [reply package.json]: Done.. and updated the Readme .. Do let me know if this is what you meant

File: sonar-project.properties (hunk 5 above)
Comment (tsukhu) [reply sonar-project.properties]: Added to the readme

File: README.md [hunk 6]
```diff
@@ -475,6 +476,39 @@ starter kit in production on [Netlify](https://www.netlify.com/):
...
+* The test results collected in the results folder in the sonar compatible format
+* Push results to SonarCube
+```bash
+sonar-scanner
+``` 
+* If working with SonarQube 6.x it supports [Generic Test Data](https://docs.sonarqube.org/display/SONAR/Generic+Test+Data)
+* Modify the [karma.conf.js](config/karma.config.js) to set the appropriate version of the sonarQube
+```json
+sonarQubeUnitReporter: {
+      sonarQubeVersion: '6.x',
```
Comment (gdi2290) [inline README.md]: fix this then I can merge it

Comment (tsukhu) [reply README.md]: Done ..

Comment (tsukhu) [reply README.md]: @gdi2290 I believe all changes are done . Let me know if anything else is required ..

**Comment by tsukhu** [comment]
@gdi2290 I have merged by changes again with the latest. Do let me know if anything else is required

**Pull request merged**